### PR TITLE
SAML feature toggles

### PR DESCRIPTION
**Pull request by arothuis**
#### Related PR
[OpenConext-deploy PR #88](https://github.com/OpenConext/OpenConext-deploy/pull/88)

#### Issues
[127988169](https://www.pivotaltracker.com/story/show/127988169)
 [127987893](https://www.pivotaltracker.com/story/show/127987893)

### CODE REVIEW

File: library/EngineBlock/Corto/Module/Bindings.php [hunk 1]
```diff
@@ -291,10 +297,21 @@ public function receiveResponse()
...
+                    'Assertions are required to be encrypted, but they are not',
+                    EngineBlock_Exception::CODE_NOTICE
+                );
+            }
+
+            if ($this->hasEncryptedAssertion($sspResponse)
+                && $this->featureConfiguration->isEnabled('eb.encrypted_assertions_require_outer_signature')
+                && !$sspResponse->isMessageConstructedWithSignature()) {
                 /** @see https://github.com/OpenConext/OpenConext-engineblock/issues/116 */
                 throw new EngineBlock_Corto_Module_Bindings_Exception(
```
Comment (arothuis) [inline library/EngineBlock/Corto/Module/Bindings.php]: Maybe change this to a custom exception?

File: library/EngineBlock/Corto/Module/Bindings.php [hunk 2]
```diff
@@ -291,10 +297,21 @@ public function receiveResponse()
             // 'Process' the response, verify the signature, verify the timings.
             $className = $this->_sspmodSamlMessageClassName;
 
-            if ($this->hasEncryptedAssertion($sspResponse) && !$sspResponse->isMessageConstructedWithSignature()) {
+            if ($this->featureConfiguration->isEnabled('eb.encrypted_assertions')
+                && !$this->hasEncryptedAssertion($sspResponse)) {
+
+                throw new EngineBlock_Corto_Module_Bindings_Exception(
```
Comment (arothuis) [inline library/EngineBlock/Corto/Module/Bindings.php]: Maybe change this to a custom exception?

File: library/EngineBlock/Corto/Module/Bindings.php (hunk 1 above)
Comment (arothuis) [inline library/EngineBlock/Corto/Module/Bindings.php]: Maybe change this to a custom exception?

File: library/EngineBlock/Corto/Module/Bindings.php [hunk 3]
```diff
@@ -291,10 +297,21 @@ public function receiveResponse()
...
+                );
+            }
+
+            if ($this->hasEncryptedAssertion($sspResponse)
+                && $this->featureConfiguration->isEnabled('eb.encrypted_assertions_require_outer_signature')
+                && !$sspResponse->isMessageConstructedWithSignature()) {
                 /** @see https://github.com/OpenConext/OpenConext-engineblock/issues/116 */
                 throw new EngineBlock_Corto_Module_Bindings_Exception(
-                    'Response signing required for use with encrypted assertions.',
+                    'Encrypted assertions are required to have an outer signature, but they have none',
```
Comment (arothuis) [inline library/EngineBlock/Corto/Module/Bindings.php]: Maybe change this to a custom exception?

**Comment by arothuis** [comment]
WIP as this is unsufficiently tested.

We should add some infrastructure for easing the automated testing of these and other feature toggles. This would require some care and work, but similar infrastructure has been built for configuring IdPs and SPs for testing purposes.

**Comment by arothuis** [comment]
@DRvanR

File: src/OpenConext/EngineBlockFunctionalTestingBundle/Features/Encryption.feature [hunk 4]
```diff
@@ -21,20 +23,65 @@ Feature:
...
+      And the IdP encrypts its assertions with the public key in "/etc/openconext/engineblock.crt"
+     When I log in at "Dummy SP"
+      And I pass through the SP
+      And I pass through EngineBlock
+      And I pass through the IdP
+     Then the url should match "authentication/feedback/received-invalid-response"
+      And I should see "Invalid Identity Provider response"
+
+  @WIP
+  Scenario: EngineBlock rejects encrypted responses without outer signature if the feature "eb.encrypted_assertions_require_outer_signatures" is enabled
```
Comment (DRvanR) [inline src/OpenConext/EngineBlockFunctionalTestingBundle/Features/Encryption.feature]: is there a positive check of this? Encrypted + outer signature?

**Comment by DRvanR** [comment]
one question, bar that, :+1:

**Comment by arothuis** [comment]
@DRvanR

**Comment by DRvanR** [comment]
:+1:

**Pull request merged**